aquacomputer_d5next.c

```diff
@@ -100,195 +133,805 @@ static const char *const label_current[] = {
+	ret = sprintf(buf, "%d\n", d5next_percent_to_pwm(val));
+	return ret;
+}
+
+static ssize_t d5next_get_auto_temp(struct device *dev, struct device_attribute *attr, char *buf)
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	int channel = d5next_userspace_to_internal_channel(sensor_attr->nr);
+	int idx = sensor_attr->index;
+	struct d5next_data *priv = dev_get_drvdata(dev);
+	int ret;
+	const char fmt[] = "%d0\n"; /* slap a 0 on the end of the reading to go from centi-degrees to milli-degrees */
+	void *to_get;
+	u16 val = 0;
+
+	if (!fan_channel_is_valid(channel))
+		return -ENOENT;
+	if (!curve_point_is_valid(idx))
+		return -ENOENT;
+	if (idx == CURVE_CONTROL_STARTING_TEMPERATURE)
+		to_get = &priv->buffer->fan_ctrl[channel].curve.start_temp;
+	else
+		to_get = &priv->buffer->fan_ctrl[channel].curve.temps[idx];
+
+	ret = d5next_get_val(dev, to_get, &val, 2);
 	if (ret < 0)
+		return ret;
+
+	ret = sprintf(buf, fmt, ntohs(val));
+	return ret;
+}
+
+static ssize_t d5next_set_auto_temp(struct device *dev, struct device_attribute *attr, const char *buf, size_t count)
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	int channel = d5next_userspace_to_internal_channel(sensor_attr->nr);
+	int idx = sensor_attr->index;
+	struct d5next_data *priv = dev_get_drvdata(dev);
+	u16 * to_set;
+	int ret;
+	long val;
+
+	if (!fan_channel_is_valid(channel))
+		return -ENOENT;
+	if (!curve_point_is_valid(idx))
+		return -ENOENT;
+	if (kstrtol(buf, 10, &val) < 0)
+		return -EINVAL;
+
+	/* pump wants values in centi-degrees Celsius, hwmon uses mdegC */
+	val = DIV_ROUND_CLOSEST(val, 10);
+
+	if (idx == CURVE_CONTROL_STARTING_TEMPERATURE)
+		to_set = &(priv->buffer->fan_ctrl[channel].curve.start_temp);
+	else
+		to_set = &(priv->buffer->fan_ctrl[channel].curve.temps[idx]);
+
+	ret = d5next_set_u16_val(dev, to_set, val);
+	return ret;
+}
+
+static ssize_t d5next_set_auto_pwm(struct device *dev, struct device_attribute *attr, const char *buf, size_t count)
+{
+	struct sensor_device_attribute_2 *sensor_attr = to_sensor_dev_attr_2(attr);
+	int channel = d5next_userspace_to_internal_channel(sensor_attr->nr);
+	int idx = sensor_attr->index;
+	struct d5next_data *priv = dev_get_drvdata(dev);
+	int ret;
+	long val;
+
+	if (!fan_channel_is_valid(channel))
+		return -ENOENT;
+	if (!curve_point_is_valid(idx))
+		return -ENOENT;
+	if (kstrtol(buf, 10, &val) < 0)
+		return -EINVAL;
+	if (val < 0 || val > 255)
+		return -EINVAL;
+
+	val = d5next_pwm_to_percent(val);
+
+	ret = d5next_set_u16_val(dev, &(priv->buffer->fan_ctrl[channel].curve.powers[idx]), val);
+	return ret;
+}
+
+#define AUTO_POINT(pwm_num, i) \
+static SENSOR_DEVICE_ATTR_2(pwm##pwm_num##_auto_point##i##_pwm, 0644, d5next_get_pwm_setting, d5next_set_auto_pwm, (pwm_num)-1, (i)-1); \
+static SENSOR_DEVICE_ATTR_2(pwm##pwm_num##_auto_point##i##_temp, 0644, d5next_get_auto_temp, d5next_set_auto_temp, (pwm_num)-1, (i)-1)
+
+AUTO_POINT(1, 1);
```

Comment:
Probably shouldn't be using macros for this. In the [docs](https://www.kernel.org/doc/html/latest/hwmon/submitting-patches.html), it says: `Avoid macros to generate groups of sensor attributes. It not only confuses checkpatch, but also makes it more difficult to review the code.`. I looked at some other drivers randomly and they have use `SENSOR_DEVICE_ATTR_2` directly.

Reply:
Oof the crashes are embarrassing! I should write some integration tests. I'm on fw 1019.

Good to know, I'll unroll the macros.

Reply:
>Reading pwm1 and pwm2 returns 0 always

When you say this, do you mean in all modes, or just curve mode? They work fine for me in manual mode. Testing curve mode now...

Reply:
Hmm, curve mode behaves as expected for me too on `pwm2`. I'm going to unroll those macros, and investigate potential FW differences.

Reply:
Curve mode

Reply:
Would you mind showing me the output of `cat pwm2_*` after you enable curve mode with `echo 2 > pwm2_enable`?

The value of `temp1_input` would be helpful as well :)

Reply:
[Here](https://gist.github.com/aleksamagicka/f8710dfc580be77a2b0a7fb839a14554) it is. Water temp is `30670`. Setting curve points doesn't crash it now :+1:

Reply:
That's interesting, because I'm not sure I actually did anything to fix the "bug". :/

Reply:
OK, I'll retest soon. I also have news from the kernel side. Recently I've sent in the Farbwerk 360 driver I've written in the meanwhile and Guenter (the hwmon maintainer) requested that it be incorporated into the existing one, so I did that and that's live in hwmon-next. So when this is completed, we'll have to rebase this (so to say) on top of that. Because of that I'll try to get the Octo driver out the door ASAP to get his feedback and minimize the final changes needed here, as it will have the basic PWM write support (so getting and sending reports).

Reply:
Are you planning to merge the octo driver into the d5/farbwerk driver as well?

Reply:
Yes, hopefully I'll send it tomorrow. Then, once it gets merged, I'll work on combining yours and that (will probably open a PR at your repo, or push here somehow if that's possible) so we can continue where we left off (I didn't really plan he'll request them to be combined, but it makes sense in hindsight)